>Integration tests (thread) — fail: 14, passed: 1469, flaky: 47

Container is not running issue. Not enough resources for such level of parallelism?

```
E               Exception: Command ['docker', 'exec', 'roottestdistributedqueriesstress_node1_r1_1', 'bash', '-c', "echo 'select * from dist_two_over_dist where key = 0;\n    select * from dist_two_over_dist where key = 1;\n    select * from dist_two_over_dist where key = 2;\n    select * from dist_two_over_dist where key = 3;\n    select * from dist_two_over_dist;' | clickhouse benchmark --concurrency=100 --cumulative --delay=0 --timelimit=3 --hedged_connection_timeout_ms=200 --connect_timeout_with_failover_ms=200 --connections_with_failover_max_tries=5 --prefer_localhost_replica=0 --optimize_skip_unused_shards=1"] return non-zero code 1: Error response from daemon: Container 39c984334f7280985a2fdfae8d7275c6e137f40bc572cda67336e5aa83bc61e9 is not running
```

```
2021.07.05 03:11:57.466621 [ 1 ] {} <Fatal> Application: Child process was terminated by signal 9 (KILL). If it is not done by 'forcestop' command or manually, the possible cause is OOM Kille
r (see 'dmesg' and look at the '/var/log/kern.log' for the details)
```

And AFAICS this test (`test_distributed_queries_stress`) hadn't passed, and it should be marked as FAIL?
- `integration_run_parallel2_1.txt` and `integration_run_parallel2_2.txt` contains only error - `runner: error: argument -t/--tests_list: expected at least one argument`
- and the row with `OK` contains failed test.

--- a/src/Interpreters/ClusterProxy/executeQuery.cpp
+++ b/src/Interpreters/ClusterProxy/executeQuery.cpp
@@ -149,6 +149,7 @@ void executeQuery(
 
             OptimizeShardingKeyRewriteInVisitor::Data visitor_data{
                 sharding_key_expr,
+                sharding_key_expr->getSampleBlock().getByPosition(0).type,
                 sharding_key_column_name,
                 shard_info,
                 not_optimized_cluster->getSlotToShard(),

--- a/src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp
+++ b/src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp
@@ -34,25 +34,18 @@ Field executeFunctionOnField(
 
 /// @param sharding_column_value - one of values from IN
 /// @param sharding_column_name - name of that column
-/// @param sharding_expr - expression of sharding_key for the Distributed() table
-/// @param sharding_key_column_name - name of the column for sharding_expr
-/// @param shard_info - info for the current shard (to compare shard_num with calculated)
-/// @param slots - weight -> shard mapping
 /// @return true if shard may contain such value (or it is unknown), otherwise false.
 bool shardContains(
     const Field & sharding_column_value,
     const std::string & sharding_column_name,
-    const ExpressionActionsPtr & sharding_expr,
-    const std::string & sharding_key_column_name,
-    const Cluster::ShardInfo & shard_info,
-    const Cluster::SlotToShard & slots)
+    const OptimizeShardingKeyRewriteInMatcher::Data & data)
 {
     /// NULL is not allowed in sharding key,
     /// so it should be safe to assume that shard cannot contain it.
     if (sharding_column_value.isNull())
         return false;
 
-    Field sharding_value = executeFunctionOnField(sharding_column_value, sharding_column_name, sharding_expr, sharding_key_column_name);
+    Field sharding_value = executeFunctionOnField(sharding_column_value, sharding_column_name, data.sharding_key_expr, data.sharding_key_column_name);
     /// The value from IN can be non-numeric,
     /// but in this case it should be convertible to numeric type, let's try.
     sharding_value = convertFieldToType(sharding_value, DataTypeUInt64());
@@ -61,8 +54,8 @@ bool shardContains(
         return false;
 
     UInt64 value = sharding_value.get<UInt64>();
-    const auto shard_num = slots[value % slots.size()] + 1;
-    return shard_info.shard_num == shard_num;
+    const auto shard_num = data.slots[value % data.slots.size()] + 1;
+    return data.shard_info.shard_num == shard_num;
 }
 
 }
@@ -92,10 +85,7 @@ void OptimizeShardingKeyRewriteInMatcher::visit(ASTFunction & function, Data & d
     if (!identifier)
         return;
 
-    const auto & sharding_expr = data.sharding_key_expr;
-    const auto & sharding_key_column_name = data.sharding_key_column_name;
-
-    if (!sharding_expr->getRequiredColumnsWithTypes().contains(identifier->name()))
+    if (!data.sharding_key_expr->getRequiredColumnsWithTypes().contains(identifier->name()))
         return;
 
     /// NOTE: that we should not take care about empty tuple,
@@ -107,7 +97,7 @@ void OptimizeShardingKeyRewriteInMatcher::visit(ASTFunction & function, Data & d
         std::erase_if(tuple_elements->children, [&](auto & child)
         {
             auto * literal = child->template as<ASTLiteral>();
-            return literal && !shardContains(literal->value, identifier->name(), sharding_expr, sharding_key_column_name, data.shard_info, data.slots);
+            return literal && !shardContains(literal->value, identifier->name(), data);
         });
     }
     else if (auto * tuple_literal = right->as<ASTLiteral>();
@@ -116,7 +106,7 @@ void OptimizeShardingKeyRewriteInMatcher::visit(ASTFunction & function, Data & d
         auto & tuple = tuple_literal->value.get<Tuple &>();
         std::erase_if(tuple, [&](auto & child)
         {
-            return !shardContains(child, identifier->name(), sharding_expr, sharding_key_column_name, data.shard_info, data.slots);
+            return !shardContains(child, identifier->name(), data);
         });
     }
 }

--- a/src/Interpreters/OptimizeShardingKeyRewriteInVisitor.h
+++ b/src/Interpreters/OptimizeShardingKeyRewriteInVisitor.h
@@ -25,9 +25,15 @@ struct OptimizeShardingKeyRewriteInMatcher
 
     struct Data
     {
+        /// Expression of sharding_key for the Distributed() table
         const ExpressionActionsPtr & sharding_key_expr;
+        /// Type of sharding_key column.
+        const DataTypePtr & sharding_key_type;
+        /// Name of the column for sharding_expr
         const std::string & sharding_key_column_name;
+        /// Info for the current shard (to compare shard_num with calculated)
         const Cluster::ShardInfo & shard_info;
+        /// weight -> shard mapping
         const Cluster::SlotToShard & slots;
     };
 

--- a/src/Storages/StorageDistributed.cpp
+++ b/src/Storages/StorageDistributed.cpp
@@ -1093,7 +1093,7 @@ ClusterPtr StorageDistributed::skipUnusedShards(
     size_t limit = local_context->getSettingsRef().optimize_skip_unused_shards_limit;
     if (!limit || limit > SSIZE_MAX)
     {
-        throw Exception("optimize_skip_unused_shards_limit out of range (0, {}]", ErrorCodes::ARGUMENT_OUT_OF_BOUND, SSIZE_MAX);
+        throw Exception(ErrorCodes::ARGUMENT_OUT_OF_BOUND, "optimize_skip_unused_shards_limit out of range (0, {}]", SSIZE_MAX);
     }
     // To interpret limit==0 as limit is reached
     ++limit;

--- a/tests/queries/0_stateless/arcadia_skip_list.txt
+++ b/tests/queries/0_stateless/arcadia_skip_list.txt
@@ -215,6 +215,7 @@
 01747_join_view_filter_dictionary
 01748_dictionary_table_dot
 01755_client_highlight_multi_line_comment_regression
+01756_optimize_skip_unused_shards_rewrite_in
 00950_dict_get
 01683_flat_dictionary
 01681_cache_dictionary_simple_key

